src/app/core/store/shopping/products/products.reducer.ts

```diff
@@ -50,13 +50,26 @@ export const productsReducer = createReducer(
   })),
   on(loadProductSuccess, (state, action) => {
     const product = action.payload.product;
-    const oldProduct = state.entities[product.sku] || { completenessLevel: 0 };
-
-    const newProduct = { ...product };
-    if (product.completenessLevel || oldProduct?.completenessLevel) {
-      newProduct.completenessLevel = Math.max(product.completenessLevel, oldProduct.completenessLevel);
+    // fetch the current product information from the state or construct an empty product stub
+    const oldProduct = state.entities[product.sku] || {
+      completenessLevel: 0,
+      available: undefined as boolean,
+      availableStock: undefined as number,
+    };
+    // update the complete product information in the state if the new product information
+    // has a higher or equal completeness level than the current product information in the state
+    if (product.completenessLevel >= oldProduct.completenessLevel) {
+      return productAdapter.upsertOne(product as AllProductTypes, {
+        ...state,
+        failed: removeFailed(state.failed, product.sku),
+      });
     }
-
+    // always update the availability information in the state with the new product information
+    const newProduct = {
+      ...oldProduct,
+      available: product.available ?? oldProduct.available,
+      availableStock: product.availableStock ?? oldProduct.availableStock,
+    };
```

Comment:
Please put the merging logic in a separate exported function and write tests for it.
From personal experience: This is probably customized in projects, so testing will help specifying custom behavior.

Reply:
Hint: the merging function should accept two products and return the product that is upserted in the reducer. Two upsert statements with one in between the logic make all of this quite unreadable.

Reply:
Thanks for the constructive critique. I hope it is now more the way you wished it would have been?
True, it looks better and I even wrote tests. 😄